Automattic/wp-calypso
# Make email confirmation check reactive

This PR makes the `user` object reactive to email confirmation, and introduces the `EmailVerificationGate` component, to disable portions of the UI if the user is unverified.

### How to Test 

1. Signup for a new account
2. Check that the verification gate is disabling access to the UI on: The menu editor, the invite users page, and on the Settings > Import page.
3. Try resending the email from the unverified notice, make sure that works.
4. Try changing the email from the unverified notice, make sure that sends you to the appropriate user profile page.
5. On a separate window _of the same browser_, open your mailbox and click the link to verify your email address. Make sure that the `EmailVerificationGate` disappears, and the UI is reenabled. (Verifying on a different browser will make the UI take longer to update, since it depends on polling)

Test live: https://calypso.live/?branch=make/email-verification-check-reactive

#### Reactive `user`

We use a combination of polling and signaling across windows (via `localStorage`) to ensure the `user` object gets the `email_verified` property updated as soon as the user confirms their email address.

To avoid unnecessary API requests, polling is only performed by the user object when the user is _not_ verified, at least one React component is subscribed to the `'verify'` event and the window is currently visible. The polling interval is 15 seconds, which is a good compromise between freshness and hitting the API server with too many requests. Most of the time, however, the `localStorage` signaling mechanism will produce an instantaneous update in the user state. (It is triggered by the Email Confirmation success message, whenever `?verified=1` is appended to the URL.)

All components that depend on email confirmation information should subscribe to `'verify'` on mount, and unsubscribe on unmount, to avoid unnecessary polling):

```javascript
class MyComponent extends React.Component {
	// …

	componentWillMount() {
		user.on( 'verify', this.updateVerificationState );
	}

	componentWillUnmount() {
		user.off( 'verify', this.updateVerificationState );
	}
}
```

#### `EmailVerificationGate`

##### What it looks like

![](https://cldup.com/tkb_JnLG-d.png)

##### How to use

```jsx
<EmailVerificationGate>
 { /* content here */ }
</EmailVerificationGate>
```

`EmailVerificationGate` wraps the content in a semitransparent div with `pointer-events: none` and logic to disable keyboard focusing of the contents. Since it only blocks the user from accessing the UI, it is complemented with a server side check.
